crates/turborepo-lib/src/shim.rs

```diff
@@ -461,129 +460,62 @@ impl InferInfo {
-                        RepoMode::MultiPackage
-                    } else {
-                        RepoMode::SinglePackage
-                    },
-                });
-
-            // More than one potential root. See if we can stop at the first.
-            // This is a performance optimization. We could remove this case,
-            // and set the mode properly in the else and it would still work.
-            } else if current.workspace_globs.is_some() {
-                // If the closest one has workspaces then we stop there.
-                return Ok(Self {
-                    root: current.path.clone(),
-                    mode: RepoMode::MultiPackage,
-                });
-
-            // More than one potential root.
-            // Closest is not RepoMode::MultiPackage
-            // We attempt to prove that the closest is a workspace of a parent.
-            // Failing that we just choose the closest.
-            } else {
-                for ancestor_infer in check_roots {
-                    if ancestor_infer.is_workspace_root_of(&current.path) {
-                        return Ok(Self {
-                            root: ancestor_infer.path.clone(),
-                            mode: RepoMode::MultiPackage,
-                        });
-                    }
-                }
-
-                // We have eliminated RepoMode::MultiPackage as an option.
-                // We must exhaustively check before this becomes the answer.
-                return Ok(Self {
-                    root: current.path.clone(),
-                    mode: RepoMode::SinglePackage,
-                });
-            }
+impl From<InferInfo> for RepoState {
+    fn from(root: InferInfo) -> Self {
+        Self {
+            mode: root.repo_mode(),
+            root: root.path,
         }
-
-        // If we're here we didn't find a valid root.
-        Err(anyhow!("Root could not be inferred."))
     }
+}
 
-    /// Infers `RepoState` from current directory.
+impl RepoState {
+    /// Infers `RepoState` from a reference path
     ///
     /// # Arguments
     ///
-    /// * `current_dir`: Current working directory
+    /// * `reference_dir`: Turbo's invocation directory
     ///
     /// returns: Result<RepoState, Error>
     pub fn infer(reference_dir: &AbsoluteSystemPath) -> Result<Self> {
-        let potential_turbo_roots = RepoState::generate_potential_turbo_roots(reference_dir);
-        RepoState::process_potential_turbo_roots(potential_turbo_roots)
+        reference_dir
+            .ancestors()
+            .filter_map(|path| {
+                PackageJson::load(&path.join_component("package.json"))
+                    .ok()
+                    .map(|package_json| {
+                        // FIXME: We should save this package manager that we detected
+                        let workspace_globs =
+                            PackageManager::get_package_manager(path, Some(&package_json))
```

Comment:
ow come `get_package_manager()` is a static method and not a method on a `PackageJSON` type? (I can't tell from the PR if `package_json` is a type of just a path or something)

Reply:
This is copying the existing setup, but, I suspect the original reason was that we may not have a `PackageJSON` instance (it's an `Option`). 

However. The only place I can currently find where we even _might_ not have a `PackageJSON` instance is [here](https://github.com/vercel/turbo/blob/fe228924e9789c11ae6cb551674009cf90b5102c/crates/turborepo-lib/src/execution_state.rs#L47). I suspect, but haven't verified, that we can't get to this point without a `PackageJSON`.

I think this is a good flag for a future refactor, and I will add a TODO, but I want to keep this PR relatively simple due to the behavior change.